boards/p1am_100/src/lib.rs

```diff
@@ -0,0 +1,292 @@
+            PushPullOutput: Led
+        }
+    }
+    PB09 {
+        name: pb09,
+        aliases: {
+            PushPullOutput: BaseEnable,
+            AlternateB: AdcBattery
+        }
+    }
+    PB10 { name: d4 }
+    PB11 { name: d5 }
+    PB22 {
+        name: d14,
+        aliases: {
+            AlternateD: UartTx
+        }
+    }
+    PB23 {
+        name: d13,
+        aliases: {
+            AlternateD: UartRx
+        }
+    }
+);
+
+const BASE_CONTROLLER_FREQ: Hertz = Hertz(1000000);
+// FIXME: const BASE_CONTROLLER_SPI_MODE: dyn spi::Mode = &spi::MODE_2;
+
+/// Convenience for setting up the labeled SPI0 peripheral.
+/// SPI0 has the P1AM base controller connected.
+/// This powers up SERCOM1 and configures it for talking to the
+/// base controller.
+pub fn base_controller_spi(
+    clocks: &mut GenericClockController,
+    sercom1: pac::SERCOM1,
+    pm: &mut pac::PM,
+    sck: Spi0Sck,
+    mosi: Spi0Mosi,
+    miso: Spi0Miso,
+) -> sercom::v2::spi::Spi<impl spi::ValidConfig> {
+    let gclk0 = &clocks.gclk0();
+    let core_clock = &clocks.sercom1_core(&gclk0).unwrap();
+    let pads = spi::Pads::new().sclk(sck).data_in(miso).data_out(mosi);
+    spi::Config::new(pm, sercom1, pads, core_clock.freq())
+        .baud(BASE_CONTROLLER_FREQ)
+        .spi_mode(spi::MODE_2)
+        .enable()
+}
+
+/// Convenience for setting up the labelled SDA, SCL pins to
+/// operate as an I2C master running at the specified frequency.
+pub fn i2c_master<F: Into<Hertz>>(
+    clocks: &mut GenericClockController,
+    bus_speed: F,
+    sercom0: pac::SERCOM0,
+    pm: &mut pac::PM,
+    sda: Sda,
+    scl: Scl,
+    port: &mut Port,
+) -> I2CMaster0<
+    impl hal::sercom::AnyPad<Sercom = pac::SERCOM0>,
+    impl hal::sercom::AnyPad<Sercom = pac::SERCOM0>,
+> {
+    let gclk0 = clocks.gclk0();
+    I2CMaster0::new(
+        &clocks.sercom0_core(&gclk0).unwrap(),
+        bus_speed.into(),
+        sercom0,
+        pm,
+        <hal::gpio::v1::Pin<_, _>>::from(sda).into_pad(port),
+        <hal::gpio::v1::Pin<_, _>>::from(scl).into_pad(port),
+    )
+}
+
+pub type Uart = UART5<
+        hal::sercom::Sercom5Pad3<<UartRx as AnyPin>::Id>,
+    hal::sercom::Sercom5Pad2<<UartTx as AnyPin>::Id>,
+    (),
+    (),
+    >;
+
+/// Convenience for setting up the labelled RX, TX pins to
+/// operate as a UART device running at the specified baud.
+pub fn uart<F: Into<Hertz>>(
+    clocks: &mut GenericClockController,
+    baud: F,
+    sercom5: pac::SERCOM5,
+    pm: &mut pac::PM,
+    rx: UartRx,
+    tx: UartTx,
+) -> Uart {
+    let gclk0 = clocks.gclk0();
+
+    UART5::new(
+        &clocks.sercom5_core(&gclk0).unwrap(),
+        baud.into(),
+        sercom5,
+        pm,
+        (rx.into(), tx.into()),
+    )
+}
+
+#[cfg(feature = "usb")]
+pub fn usb_allocator(
+    usb: pac::USB,
+    clocks: &mut GenericClockController,
+    pm: &mut pac::PM,
+    dm: impl AnyPin<Id = <UsbDm as AnyPin>::Id>,
+    dp: impl AnyPin<Id = <UsbDp as AnyPin>::Id>,
```

Comment:
Why not use `UsbDm` and `UsbDp` here, like above? The `impl AnyPin` approach is only there to accept the same pin in any `PinMode`. But in this case, it seems like consistency would be more important.

Reply:
Good point; I wrote this first before I figured out how to use a pin alias. Fixed.